Model override from commandline

chriselion opened the pull request:
Not useful on its own yet, but I think is how we can get training outputs back into a built player to measure inference.

To use it fully, we'd need to add to the Agent prefabs in all the scenes. The workflow in testing would be something like:
1) Run training via mlagents-learn
2) Run just the executlable with `--mlagents-override-model <behavior name> <path to model>`
3) Write rewards results to file somewhere

chriselion commented on UnitySDK/Assets/ML-Agents/Scripts/Agent.cs:
```diff
@@ -292,6 +292,11 @@ void NotifyAgentDone()
+            if (m_PolicyFactory == null)
+            {
+                // TODO better way to do this lazily
+                m_PolicyFactory = GetComponent<BehaviorParameters>();
```
At least in my one test, the ModelNotifier.OnEnable was called before the Agent, so m_PolicyFactory wasn't initialized yet. Is there a better way to go about this?

surfnerd replied:
maybe the initialization of the `BehaviorParameters` (and other 'static' members) should be moved to `Awake`

vincentpierre replied:
Maybe we should create an AgentLazyInitialize to make sure it is initialized correctly?
This way we could raise proper errors if the m_PolicyFactory is not attached.

chriselion replied:
Actually, I think I can get around this for now by accessing the BehaviorParameters directly.

chriselion replied:
Ended up doing Agent.LazyInitialize(). This makes some unit test code cleaner too.

vincentpierre commented on UnitySDK/Assets/ML-Agents/Examples/SharedAssets/Scripts/ModelOverrider.cs:
```diff
@@ -0,0 +1,58 @@
+            asset.Value = model;
+            asset.name = "Override - " + Path.GetFileName(m_AssetPath);
+            
+            var agent = GetComponent<Agent>();
```
This will get the first component Agent in the scene. If it contains multiple Agents, this might not work. 
I think it won't work at all for WallJump but if it is only for our demo environments then this is fine.

chriselion replied:
Yeah, will probably need some more logic with multiple brains on the the same object, or multiple agent types in the scene (e.g. soccer). I think probably specify key-value pairs on the command line, and remap them here.

chriselion replied:
Do we ever have a GameObject with multiple agents? WallJump has 1 agent and 1 BehaviorParameters (per prefab)

chriselion replied:
Added a comment in the summary that it will only work with scenes that have 1:1 Agent:Behavior.

surfnerd replied:
It won't get the first Agent in the scene, but the first Agent on the gameobject this is also attached to.

surfnerd replied:
I don't recall if we have multiple agents on a single gameobject

vincentpierre commented on UnitySDK/Assets/ML-Agents/Examples/SharedAssets/Scripts/ModelOverrider.cs:
```diff
@@ -0,0 +1,58 @@
+using UnityEngine;
+using Barracuda;
+using System.IO;
+namespace MLAgents
```
There is some code in the repo (this code included) that is not for our users to use. I wonder if we should have a separate namespace for those.

chriselion replied:
good idea, I didn't realize they were in the same namespace. maybe `MLAgentsExamples`?

chriselion replied:
Renamed to MLAgentsExamples (except RayPerception, because I'm going to delete that soon).

chriselion commented on UnitySDK/Assets/ML-Agents/Examples/SharedAssets/Scripts/ModelOverrider.cs:
```diff
@@ -0,0 +1,110 @@
+            agent.LazyInitialize();
+            var bp = agent.GetComponent<BehaviorParameters>();
+
+            var behaviorNameAndTeamId = bp.behaviorName;
```
Going to revisit this later, but I think we need some better accessors on BehaviorParameters.

chriselion commented on UnitySDK/Assets/ML-Agents/Examples/SharedAssets/Scripts/ProjectSettingsOverrides.cs:
```diff
@@ -1,60 +1,63 @@
-        Physics.defaultSolverVelocityIterations = solverVelocityIterations;
-
-        Academy.Instance.FloatProperties.RegisterCallback("gravity", f => { Physics.gravity = new Vector3(0, -f, 0); });
+        // Original values
```
Forgot the namespace on this one originally 🤦‍♂

chriselion commented:
Changes from initial version:
* Added MLAgentsExamples namespace
* Added Agent.LazyInitialize()
* Updated the ModelOverrider so that it takes (behaviorName, asset path) pairs on the command line.

surfnerd commented on UnitySDK/Assets/ML-Agents/Examples/FoodCollector/Scripts/FoodCollectorArea.cs:
```diff
@@ -1,5 +1,5 @@
 using UnityEngine;
-using MLAgents;
+using MLAgentsExamples;
```
👍

vincentpierre commented on UnitySDK/Assets/ML-Agents/Scripts/Agent.cs:
```diff
@@ -221,16 +224,25 @@ public AgentInfo Info
         /// Helper method for the <see cref="OnEnable"/> event, created to
         /// facilitate testing.
-        void OnEnableHelper()
+        public void LazyInitialize()
```
Should this really be public ?

chriselion replied:
It needs to be public - internal won't cut it since the ModelOverrider calls it, and that's in the MLAgentsExamples namespace.
